psycopg/psycopg/waiting.py

```diff
@@ -110,38 +110,29 @@ async def wait_async(gen: PQGen[RV], fileno: int) -> RV:
 
     Behave like in `wait()`, but exposing an `asyncio` interface.
     """
-    # Use an event to block and restart after the fd state changes.
-    # Not sure this is the best implementation but it's a start.
-    ev = Event()
     loop = get_event_loop()
-    ready: Ready
     s: Wait
 
-    def wakeup(state: Ready) -> None:
-        nonlocal ready
-        ready = state
-        ev.set()
-
     try:
         s = next(gen)
         while 1:
-            ev.clear()
-            if s == Wait.R:
-                loop.add_reader(fileno, wakeup, Ready.R)
-                await ev.wait()
-                loop.remove_reader(fileno)
-            elif s == Wait.W:
-                loop.add_writer(fileno, wakeup, Ready.W)
-                await ev.wait()
-                loop.remove_writer(fileno)
-            elif s == Wait.RW:
-                loop.add_reader(fileno, wakeup, Ready.R)
-                loop.add_writer(fileno, wakeup, Ready.W)
-                await ev.wait()
-                loop.remove_reader(fileno)
-                loop.remove_writer(fileno)
-            else:
-                raise e.InternalError("bad poll status: %s")
+            fs = []
+            if s & Wait.R:
+                rf = loop.create_future()
+                loop.add_reader(fileno, rf.set_result, Ready.R)
+                rf.add_done_callback(lambda f: loop.remove_reader(fileno))
+                fs.append(rf)
+            if s & Wait.W:
+                wf = loop.create_future()
+                loop.add_writer(fileno, wf.set_result, Ready.W)
+                wf.add_done_callback(lambda f: loop.remove_writer(fileno))
+                fs.append(wf)
+            if not fs:
+                raise e.InternalError(f"bad poll status: {s}")
+            done, _ = await asyncio_wait(fs, return_when=FIRST_COMPLETED)
```

Comment:
use `[0]` instead of unpacking the tuple and discarding the element

Reply:
No longer applies as we've moved back to Event-based implementation.